src/Select.js

```diff
@@ -569,30 +569,60 @@ var Select = React.createClass({
 			}
 		}
 
-		this.props.asyncOptions(input, (err, data) => {
-			if (err) throw err;
-			if (this.props.cacheAsyncResults) {
-				this._optionsCache[input] = data;
-			}
-			if (thisRequestId !== this._currentRequestId) {
-				return;
-			}
-			var filteredOptions = this.filterOptions(data.options);
-			var newState = {
-				options: data.options,
-				filteredOptions: filteredOptions,
-				focusedOption: this._getNewFocusedOption(filteredOptions)
-			};
-			for (var key in state) {
-				if (state.hasOwnProperty(key)) {
-					newState[key] = state[key];
+		if (typeof this.props.asyncOptions() !== 'undefined' &&
+			typeof this.props.asyncOptions().then === 'function') {
+			this.props.asyncOptions(input).then((data) => {
```

Comment:
You call `this.props.asyncOptions(input)` and assign the result to a variable. Then check the variable instead of subsequent calls to `this.props.asyncOptions()`.

Basically:

```
		let result = this.props.asyncOptions(input);
		if (result !== 'undefined' &&
			typeof result.then === 'function') {
			result.then((data) => {
```